# Data Cleaning and Preperation
Imports supplied data, inspects their structure and missing values. Standardise cluster identifiers and prepares the datasets for upstream analysis

### Import packages
Load the pandas package and let me refer to it using the shorter name "pd" (working with tables/data). 
   
Load the numpy package and let me refer to it using the shorter name "np" (numerical calculations).  

In [1]:
import pandas as pd
import numpy as np

### Reading the files
`variable_name = pd.read_csv("../data/raw/file_name.csv)` - creates a varaible called `variable_name` - that stores the contents of `file_name` inside it.   
`pd.read_csv` - `pd` refers to pandas package, `read_csv` pandas function for opening csv files.  
`../` - tells Python to go outside the current folder into data then raw to find the csv file.  

In [2]:
vdb = pd.read_csv("../data/raw/vandenBerg_table2.csv")
krause = pd.read_csv("../data/raw/Krause21.csv")
harris1 = pd.read_csv("../data/raw/HarrisPartI.csv")
harris3 = pd.read_csv("../data/raw/HarrisPartIII.csv")

## Finding size of each dataset 
`.shape` means: (number of rows, number of columns)

In [3]:
print("VandenBerg:", vdb.shape)
print("Krause:", krause.shape)
print("Harris Part I:", harris1.shape)
print("Harris Part III:", harris3.shape)

VandenBerg: (55, 13)
Krause: (61, 8)
Harris Part I: (157, 11)
Harris Part III: (157, 13)


Therefore, VandenBerg has 55 globular clusters and 13 variables/columns.

## Finding column names 

In [4]:
print(vdb.columns)
print (krause.columns)
print (harris1.columns)
print (harris3.columns)

Index(['#NGC', 'Name', 'FeH', 'Age', 'Age_err', 'Method', 'Figs', 'Range',
       'HBtype', 'R_G', 'M_V', 'v_e0', 'log_sigma_0'],
      dtype='str')
Index(['Class', 'Object', 'AltName', 'Mstar', 'rh', 'C5', 'Age', 'FeH'], dtype='str')
Index(['ID', 'Name', 'RA', 'DEC', 'L', 'B', 'R_Sun', 'R_gc', 'X', 'Y', 'Z'], dtype='str')
Index(['ID', 'v_r', 'v_r_e', 'v_LSR', 'sig_v', 'sig_v_e', 'c', 'r_c', 'r_h',
       'mu_V', 'rho_0', 'lg_tc', 'lg_th'],
      dtype='str')


## Generating Table 
`head()` - outputs the first 5 rows. 

In [5]:
vdb.head()

,#NGC,Name,FeH,Age,Age_err,Method,Figs,Range,HBtype,R_G,M_V,v_e0,log_sigma_0
0,104,47Tuc,-0.76,11.75,0.25,V,14,11.50–11.75,-0.99,7.4,-9.42,54.8,5.061
1,288,XXXX,-1.32,11.50,0.38,H,24,NaN,0.98,12.0,-6.75,10.9,2.953
2,362,XXXX,-1.30,10.75,0.25,V,13,10.75–11.00,-0.87,9.4,-8.43,44.4,4.938
3,1261,XXXX,-1.27,10.75,0.25,V,13,10.75–11.25,-0.71,18.1,-7.80,23.6,3.913
4,1851,XXXX,-1.18,11.00,0.25,V,13,10.75–11.25,-0.32,16.6,-8.33,47.6,5.136


In [6]:
krause.head()

,Class,Object,AltName,Mstar,rh,C5,Age,FeH
0,GC,NGC104,47Tuc,6.46,7.10,0.92,12.80,-0.76
1,GC,NGC288,Melotte3,0.46,9.80,0.05,12.20,-1.32
2,GC,NGC362,Dunlop62,2.50,3.50,0.72,10.00,-1.26
3,GC,NGC1261,Caldwell87,3.41,5.50,0.62,10.24,-1.08
4,GC,NGC1851,Dunlop508,5.51,3.05,1.81,7.64,-1.13


In [7]:
harris1.head()

,ID,Name,RA,DEC,L,B,R_Sun,R_gc,X,Y,Z
0,NGC 104,47 Tuc,00:24:05.67,-72:04:52.6,305.89,-44.89,4.5,7.4,1.9,-2.6,-3.1
1,NGC 288,NaN,00:52:45.24,-26:34:57.4,152.30,-89.38,8.9,12.0,-0.1,0.0,-8.9
2,NGC 362,NaN,01:03:14.26,-70:50:55.6,301.53,-46.25,8.6,9.4,3.1,-5.1,-6.2
3,Whiting 1,NaN,02:02:57,-03:15:10,161.22,-60.76,30.1,34.5,-13.9,4.7,-26.3
4,NGC 1261,NaN,03:12:16.21,-55:12:58.4,270.54,-52.12,16.3,18.1,0.1,-10.0,-12.9


In [8]:
harris3.head()

,ID,v_r,v_r_e,v_LSR,sig_v,sig_v_e,c,r_c,r_h,mu_V,rho_0,lg_tc,lg_th
0,NGC 104,-18.0,0.1,-26.7,11.0,0.3,2.07,0.36,3.17,14.38,4.88,7.84,9.55
1,NGC 288,-45.4,0.2,-51.9,2.9,0.3,0.99,1.35,2.23,20.05,1.78,8.99,9.32
2,NGC 362,223.5,0.5,214.0,6.4,0.3,1.76,0.18,0.82,14.80,4.74,7.76,8.93
3,Whiting 1,-130.6,1.8,-138.1,NaN,NaN,0.55,0.25,0.22,NaN,NaN,NaN,7.93
4,NGC 1261,68.2,4.6,54.9,NaN,NaN,1.16,0.35,0.68,17.73,2.99,8.59,9.12


## Checking for missing values 
Missing values were checked in each catalogue before data cleaning and merging. No rows were removed at this stage.
 
`vdb.insa().sum()` - 0 indicates no missing values, >0 indicates that number of missing values.  
`.insa()` - asks Python, "is this value missing".  
`.sum()` - counts how many missing values in each column.  

In [9]:
vdb.isna().sum()

#NGC            0
Name            0
FeH             0
Age             0
Age_err         0
Method          0
Figs            0
Range          22
HBtype          0
R_G             0
M_V             0
v_e0            0
log_sigma_0     0
dtype: int64

In [10]:
krause.isna().sum()

Class      0
Object     0
AltName    0
Mstar      0
rh         0
C5         0
Age        0
FeH        0
dtype: int64

In [12]:
harris1.isna().sum()

ID         0
Name     106
RA         0
DEC        0
L          0
B          0
R_Sun      0
R_gc       0
X          0
Y          0
Z          0
dtype: int64

In [14]:
harris3.isna().sum()

ID          0
v_r        14
v_r_e      14
v_LSR      14
sig_v      95
sig_v_e    95
c           1
r_c         1
r_h         5
mu_V       14
rho_0      14
lg_tc      14
lg_th       2
dtype: int64

### Missing Data observations
- VandenBerg contains 22 missing values in Range, but the key age and metallicity variables are complete
- Krause21 contains no missing values. 
- Harris Part I contains 106 missing entries in the optional Name column, while ID and positional information is complete. 
- Harris Part III contains missing kinematic and stricutral measurements, particularly sig_v. These rows will be retained initally and handled according to the varaiables used in later analysis. 

## Standardising Cluster Identifiers
Cluster identiiferiers are formatted differently between catalouges. A common Cluster_ID column will be created to allow the datasets to be matched and merged. 

### VandenBerg Cluster Identifiers
Most clusters are identified by an NGC number. For clusters without an NGC number (#NGC = XXXX), the listed cluster name is used instead.

`[["#NGC", "Name"]]` - only shows me the columns called #NGC and Name.  
`.tail(10)` - shows me the last 10 rows.  

We only want the last 10 rows of the VandenBerg table as the unusal non-NGC cluster entries are near the bottom. 

In [23]:
print(vdb[["#NGC", "Name"]].tail(10))

    #NGC   Name
45  6809    M55
46  6838    M71
47  6934   XXXX
48  6981    M72
49  7078    M15
50  7089     M2
51  7099    M30
52  XXXX  Arp21
53  XXXX  Pal12
54  XXXX   Ter8


`vbd["Cluster_ID"] =` - creates a new column in the VandenBerg table called Cluster_ID
`np.where(condition, do_this_if_true, do_this_if_false)` - works like an if/else statement

`vdb["#NGC"].astype(str).str.strip() != "XXXX"` - Look at the #NGC value, treat it as text, remove any accidental spaces, and ask whether it is not equal to XXXX.

If thats true then, `"NGC" + vdb["#NGC"].astype(str).str.strip()`.  
If its false (meaning NGC = XXXX) then, `vdb["Name"].astype(str).str.replace(" ", "", regex=False)`.  

`.astype(str)` - convert data to str, a string. 
`.str.strip()` - removes spaces from the beginning and end of text
`.str.replace(" ", "", regex=False)` - removes spaces inside of text, " " gets replaced with "". The regex=False treats the thing we are replacing as normal text, not as a special pattern. 

In [24]:
vdb["Cluster_ID"] = np.where(
    vdb["#NGC"].astype(str).str.strip() != "XXXX",
    "NGC" + vdb["#NGC"].astype(str).str.strip(),
    vdb["Name"].astype(str).str.replace(" ", "", regex=False)
)

In [25]:
print(vdb[["#NGC", "Name", "Cluster_ID"]].tail(10))

    #NGC   Name Cluster_ID
45  6809    M55    NGC6809
46  6838    M71    NGC6838
47  6934   XXXX    NGC6934
48  6981    M72    NGC6981
49  7078    M15    NGC7078
50  7089     M2    NGC7089
51  7099    M30    NGC7099
52  XXXX  Arp21      Arp21
53  XXXX  Pal12      Pal12
54  XXXX   Ter8       Ter8


### Harris Cluster Identifiers
Similarly, spaces were removed from the Harris catalogue identifiers to create a consistent Cluster_ID format for catalogue matching.

In [26]:
harris1["Cluster_ID"] = (
    harris1["ID"]
    .astype(str)
    .str.strip()
    .str.replace(" ", "", regex=False)
)

harris3["Cluster_ID"] = (
    harris3["ID"]
    .astype(str)
    .str.strip()
    .str.replace(" ", "", regex=False)
)

Checking that this worked:

In [27]:
print(harris1[["ID", "Cluster_ID"]].head())
print(harris1[["ID", "Cluster_ID"]].tail())

          ID Cluster_ID
0    NGC 104     NGC104
1    NGC 288     NGC288
2    NGC 362     NGC362
3  Whiting 1   Whiting1
4   NGC 1261    NGC1261
           ID Cluster_ID
152  NGC 7089    NGC7089
153  NGC 7099    NGC7099
154    Pal 12      Pal12
155    Pal 13      Pal13
156  NGC 7492    NGC7492


### Krause Cluster Identifiers
A standardised Cluster_ID column was also created for the Krause catalogue.

In [28]:
krause["Cluster_ID"] = (
    krause["Object"]
    .astype(str)
    .str.strip()
    .str.replace(" ", "", regex=False)
)

print(krause[["Object", "Cluster_ID"]].head())
print(krause[["Object", "Cluster_ID"]].tail())

    Object Cluster_ID
0   NGC104     NGC104
1   NGC288     NGC288
2   NGC362     NGC362
3  NGC1261    NGC1261
4  NGC1851    NGC1851
         Object   Cluster_ID
56      NGC7089      NGC7089
57      NGC7099      NGC7099
58  Ruprecht106  Ruprecht106
59      Terzan7      Terzan7
60    Palomar12    Palomar12


### Checking Catalouge Matches
The standardised cluster identifiers were compared between catalouges to identify clusters that do not match automatically, for example Pal12 ≠ Palomar12 but they refer to the same cluster. 

`vbd["Cluster_ID"]` - takes the cleaned ID column.  
`set(...)` - turns that column into a collection of unique values.  
`vdb_ids*/*harris1_ids` - names given to those two collections.  

In [45]:
vdb_ids = set(vdb["Cluster_ID"])
harris1_ids = set(harris1["Cluster_ID"])

By *-* these collections it gives us everthing that exists in *vbd_ids* but not *harris_ids*

In [46]:
vdb_not_in_harris1 = vdb_ids - harris1_ids

print(vdb_not_in_harris1)

{'Arp21', 'Ter8'}


Doing the same for Harris Part III:

In [47]:
harris3_ids = set(harris3["Cluster_ID"])

vdb_not_in_harris3 = vdb_ids - harris3_ids

print(vdb_not_in_harris3)

{'Arp21', 'Ter8'}


Check between Harris Part I and Harris Part III:

In [48]:
print(harris1_ids == harris3_ids)

True


53 of the 55 VandenBerg clusters matched directly to both Harris catalogues after standardising identifier formatting.   
Arp21 and Ter8 did not match automatically and require manual investigation of alternative naming.

Doing the same for Krause IDS: 

In [51]:
krause_ids = set(krause["Cluster_ID"])

vdb_not_in_krause = vdb_ids - krause_ids
print("VandenBerg IDs not in Krause:")
print(vdb_not_in_krause)

VandenBerg IDs not in Krause:
{'Arp21', 'Ter8', 'NGC4147', 'Pal12'}


In [52]:
krause_not_in_vdb = krause_ids - vdb_ids
print("Krause IDs not in VandenBerg:")
print(krause_not_in_vdb)

Krause IDs not in VandenBerg:
{'NGC6093', 'Palomar12', 'NGC6388', 'NGC6139', 'NGC6441', 'NGC1904', 'NGC5897', 'NGC2298', 'Terzan7', 'Ruprecht106'}


### Resolving Catalouge Naming Differences
Some clusters use different names or abbreviations between catalogues. Verified alternative names were mapped to a common identifier. Objects genuinely absent from another catalogue were not altered.

`"old value":"new value"`.  
For example, `"Ter8": "Terzan8"` means whenever we find Ter8, replace it with Terzan8.

In [53]:
vdb_aliases = {
    "Ter8": "Terzan8"
}

krause_aliases = {
    "Palomar12": "Pal12",
    "Ruprecht106": "Rup106"
}

vdb["Cluster_ID"] = vdb["Cluster_ID"].replace(vdb_aliases)

krause["Cluster_ID"] = krause["Cluster_ID"].replace(krause_aliases)

Checking the sets again after changing the IDs: 

In [54]:
vdb_ids = set(vdb["Cluster_ID"])
krause_ids = set(krause["Cluster_ID"])
harris1_ids = set(harris1["Cluster_ID"])

print("VandenBerg not in Harris:")
print(vdb_ids - harris1_ids)

print("\nVandenBerg not in Krause:")
print(vdb_ids - krause_ids)

print("\nKrause not in Harris:")
print(krause_ids - harris1_ids)

VandenBerg not in Harris:
{'Arp21'}

VandenBerg not in Krause:
{'Arp21', 'Terzan8', 'NGC4147'}

Krause not in Harris:
set()


Arp21 in VandenBerg catalogue does not match the Harris catalogue and need to verify it against the original source. 

### Mannual Identifier Corrections
The original VandenBerg et al. (2013) Table 2 was checked to resolve catalogue naming discrepancies. The supplied Arp21 identifier corresponds to Arp 2, and Ter8 corresponds to Terzan 8 in the Harris catalogue naming convention.

In [55]:
vdb_aliases = {
    "Arp21": "Arp2",
    "Ter8": "Terzan8"
}

vdb["Cluster_ID"] = vdb["Cluster_ID"].replace(vdb_aliases)

krause_aliases = {
    "Palomar12": "Pal12",
    "Ruprecht106": "Rup106"
}

krause["Cluster_ID"] = krause["Cluster_ID"].replace(krause_aliases)

Re-checking the matches: 

In [56]:
vdb_ids = set(vdb["Cluster_ID"])
krause_ids = set(krause["Cluster_ID"])
harris1_ids = set(harris1["Cluster_ID"])
harris3_ids = set(harris3["Cluster_ID"])

print("VandenBerg not in Harris I:")
print(vdb_ids - harris1_ids)

print("\nVandenBerg not in Harris III:")
print(vdb_ids - harris3_ids)

VandenBerg not in Harris I:
set()

VandenBerg not in Harris III:
set()


Now every VandenBerg cluster has a matching Harris cluster. 

## Merging Catalogues
### Merging Harris Part I and Part III
`pd.merge(...)` - use pandas' table merging function.  
`on="Cluster_ID"` - use Cluster_ID to decide which rows belong together.  
`how=inner` - only keeps clusters that occour in both tables.  
`suffixes=("_pos","_kin")` - if both tables contain a column with the same name, pandas distinguishes them adding suffixes.  
`validate="one_to_one"` - tells pandas each cluster should appear once. If not, Python will not merge. 

In [57]:
harris = pd.merge(
    harris1,
    harris3,
    on="Cluster_ID",
    how="inner",
    suffixes=("_pos", "_kin"),
    validate="one_to_one"
)

To check the merge worked, we can check the shape since both original Harris tables had 157 clusters we should still have 157 rows. 

In [58]:
print(harris.shape)

(157, 25)


In [59]:
harris.head()

,ID_pos,Name,RA,DEC,L,B,R_Sun,R_gc,X,Y,...,v_LSR,sig_v,sig_v_e,c,r_c,r_h,mu_V,rho_0,lg_tc,lg_th
0,NGC 104,47 Tuc,00:24:05.67,-72:04:52.6,305.89,-44.89,4.5,7.4,1.9,-2.6,...,-26.7,11.0,0.3,2.07,0.36,3.17,14.38,4.88,7.84,9.55
1,NGC 288,NaN,00:52:45.24,-26:34:57.4,152.30,-89.38,8.9,12.0,-0.1,0.0,...,-51.9,2.9,0.3,0.99,1.35,2.23,20.05,1.78,8.99,9.32
2,NGC 362,NaN,01:03:14.26,-70:50:55.6,301.53,-46.25,8.6,9.4,3.1,-5.1,...,214.0,6.4,0.3,1.76,0.18,0.82,14.80,4.74,7.76,8.93
3,Whiting 1,NaN,02:02:57,-03:15:10,161.22,-60.76,30.1,34.5,-13.9,4.7,...,-138.1,NaN,NaN,0.55,0.25,0.22,NaN,NaN,NaN,7.93
4,NGC 1261,NaN,03:12:16.21,-55:12:58.4,270.54,-52.12,16.3,18.1,0.1,-10.0,...,54.9,NaN,NaN,1.16,0.35,0.68,17.73,2.99,8.59,9.12


`.nunique()` - number of unique values.

In [60]:
print(harris["Cluster_ID"].nunique())

157


### Merging VandenBerg onto Harris
As VandenBerg has 55 clusters compared to Harris 157 and the project is mainly based around the 55 VandenBerg clustres with age/metallicity measurements we can keep all the VandenBerg rows and add Harris whenever a matching cluster exists. 

`how=left` - keeps every row from the left-hand table (VandenBerg) and adds the matching information from the right-hand table wherever possible. 

In [61]:
master = pd.merge(
    vdb,
    harris,
    on="Cluster_ID",
    how="left",
    validate="one_to_one"
)

Checking this:

In [62]:
print(master.shape)
master.head()

(55, 38)


,#NGC,Name_x,FeH,Age,Age_err,Method,Figs,Range,HBtype,R_G,...,v_LSR,sig_v,sig_v_e,c,r_c,r_h,mu_V,rho_0,lg_tc,lg_th
0,104,47Tuc,-0.76,11.75,0.25,V,14,11.50–11.75,-0.99,7.4,...,-26.7,11.0,0.3,2.07,0.36,3.17,14.38,4.88,7.84,9.55
1,288,XXXX,-1.32,11.50,0.38,H,24,NaN,0.98,12.0,...,-51.9,2.9,0.3,0.99,1.35,2.23,20.05,1.78,8.99,9.32
2,362,XXXX,-1.30,10.75,0.25,V,13,10.75–11.00,-0.87,9.4,...,214.0,6.4,0.3,1.76,0.18,0.82,14.80,4.74,7.76,8.93
3,1261,XXXX,-1.27,10.75,0.25,V,13,10.75–11.25,-0.71,18.1,...,54.9,NaN,NaN,1.16,0.35,0.68,17.73,2.99,8.59,9.12
4,1851,XXXX,-1.18,11.00,0.25,V,13,10.75–11.25,-0.32,16.6,...,303.8,10.4,0.5,1.86,0.09,0.51,14.25,5.09,7.43,8.82


In [63]:
print(master["Cluster_ID"].nunique())

55


Check if the following variables are in the master: 

In [64]:
master[
    ["Cluster_ID", "Age", "Age_err", "FeH", "R_gc", "v_r", "v_LSR"]
].head()

,Cluster_ID,Age,Age_err,FeH,R_gc,v_r,v_LSR
0,NGC104,11.75,0.25,-0.76,7.4,-18.0,-26.7
1,NGC288,11.50,0.38,-1.32,12.0,-45.4,-51.9
2,NGC362,10.75,0.25,-1.30,9.4,223.5,214.0
3,NGC1261,10.75,0.25,-1.27,18.1,68.2,54.9
4,NGC1851,11.00,0.25,-1.18,16.6,320.5,303.8


Check the missing values again but only in the important columns: 

In [65]:
important_columns = [
    "Age",
    "Age_err",
    "FeH",
    "R_gc",
    "v_r",
    "v_r_e",
    "v_LSR",
    "sig_v",
    "sig_v_e"
]

master[important_columns].isna().sum()

Age         0
Age_err     0
FeH         0
R_gc        0
v_r         0
v_r_e       0
v_LSR       0
sig_v      18
sig_v_e    18
dtype: int64

All 55 VandenBerg clusters matched to the Harris catalogues. Age, metallicity, Galactocentric distance, radial velocity and LSR velocity are complete for the merged sample. Velocity dispersion measurements are unavailable for 18 clusters.

### Merging Krause onto master
Renaming krause column names common to master:

In [69]:
krause = krause.rename(columns={
    "Age": "Age_Krause",
    "FeH": "FeH_Krause"
})

Merge Krause onto master while keeping VandenBerg as the primary dataset:

In [70]:
master = pd.merge(
    master,
    krause[
        ["Cluster_ID", "Mstar", "rh", "C5", "Age_Krause", "FeH_Krause"]
    ],
    on="Cluster_ID",
    how="left",
    validate="one_to_one"
)

Check:

In [ ]:
print(master.shape)
print(master["Cluster_ID"].nunique())

(55, 43)
55


Mstar         3
rh            3
C5            3
Age_Krause    3
FeH_Krause    3
dtype: int64

In [72]:
master[
    ["Mstar", "rh", "C5", "Age_Krause", "FeH_Krause"]
].isna().sum()

Mstar         3
rh            3
C5            3
Age_Krause    3
FeH_Krause    3
dtype: int64

All 5 Krause varaibles are missing for exactly three rows, it suggests that those three clusteers aren't in the Krause catalogue rather than measurements randomly being missing. Finding out which clusters these are:

In [73]:
master[
    master["Mstar"].isna()
][["Cluster_ID", "Age", "FeH"]]

,Cluster_ID,Age,FeH
7,NGC4147,12.25,-1.78
52,Arp2,12.00,-1.74
54,Terzan8,13.00,-2.34


Krause measurements were successfully matched to 52 of the 55 clusters in the primary VandenBerg sample. Three clusters are absent from the Krause catalogue and retain missing values for the Krause-derived mass, radius, compactness, age and metallicity parameters.

## Removing redundant columns
`columns.tolist()` - turns the columns into a normal Python list

In [75]:
print(master.columns.tolist())

['#NGC', 'Name_x', 'FeH', 'Age', 'Age_err', 'Method', 'Figs', 'Range', 'HBtype', 'R_G', 'M_V', 'v_e0', 'log_sigma_0', 'Cluster_ID', 'ID_pos', 'Name_y', 'RA', 'DEC', 'L', 'B', 'R_Sun', 'R_gc', 'X', 'Y', 'Z', 'ID_kin', 'v_r', 'v_r_e', 'v_LSR', 'sig_v', 'sig_v_e', 'c', 'r_c', 'r_h', 'mu_V', 'rho_0', 'lg_tc', 'lg_th', 'Mstar', 'rh', 'C5', 'Age_Krause', 'FeH_Krause']


Need to remove old identifier columns that were only needed to build Cluster_ID. 

In [77]:
cleaned = master.drop(columns=[
    "#NGC",
    "Name_x",
    "ID_pos",
    "Name_y",
    "ID_kin"
])

print(cleaned.shape)
print(cleaned.columns.tolist())

(55, 38)
['FeH', 'Age', 'Age_err', 'Method', 'Figs', 'Range', 'HBtype', 'R_G', 'M_V', 'v_e0', 'log_sigma_0', 'Cluster_ID', 'RA', 'DEC', 'L', 'B', 'R_Sun', 'R_gc', 'X', 'Y', 'Z', 'v_r', 'v_r_e', 'v_LSR', 'sig_v', 'sig_v_e', 'c', 'r_c', 'r_h', 'mu_V', 'rho_0', 'lg_tc', 'lg_th', 'Mstar', 'rh', 'C5', 'Age_Krause', 'FeH_Krause']


Add suffixes to duplicate varaibles. For example, the r_h from Harris and rh from Krause:

In [ ]:
cleaned = cleaned.rename(columns={
    "R_G": "R_G_VDB",
    "R_gc": "R_gc_Harris",
    "r_h": "r_h_Harris",
    "rh": "rh_Krause"
})

Check:

In [78]:
print(cleaned.shape)
print(cleaned["Cluster_ID"].nunique())
cleaned.head()

(55, 38)
55


,FeH,Age,Age_err,Method,Figs,Range,HBtype,R_G,M_V,v_e0,...,r_h,mu_V,rho_0,lg_tc,lg_th,Mstar,rh,C5,Age_Krause,FeH_Krause
0,-0.76,11.75,0.25,V,14,11.50–11.75,-0.99,7.4,-9.42,54.8,...,3.17,14.38,4.88,7.84,9.55,6.46,7.10,0.92,12.80,-0.76
1,-1.32,11.50,0.38,H,24,NaN,0.98,12.0,-6.75,10.9,...,2.23,20.05,1.78,8.99,9.32,0.46,9.80,0.05,12.20,-1.32
2,-1.30,10.75,0.25,V,13,10.75–11.00,-0.87,9.4,-8.43,44.4,...,0.82,14.80,4.74,7.76,8.93,2.50,3.50,0.72,10.00,-1.26
3,-1.27,10.75,0.25,V,13,10.75–11.25,-0.71,18.1,-7.80,23.6,...,0.68,17.73,2.99,8.59,9.12,3.41,5.50,0.62,10.24,-1.08
4,-1.18,11.00,0.25,V,13,10.75–11.25,-0.32,16.6,-8.33,47.6,...,0.51,14.25,5.09,7.43,8.82,5.51,3.05,1.81,7.64,-1.13


Saving this cleaned data:

In [80]:
cleaned.to_csv(
    "../data/cleaned/globular_clusters_cleaned.csv",
    index=False
)

## Data Dictionary

The cleaned dataset contains 55 globular clusters from the VandenBerg sample, with additional positional and kinematic information from the Harris catalogue and structural information from Krause et al.

| Column | Description |
|---|---|
| `Cluster_ID` | Standardised globular cluster identifier used to match the catalogues |
| `FeH` | Metallicity, [Fe/H], from VandenBerg |
| `Age` | Cluster age from VandenBerg |
| `Age_err` | Uncertainty in the VandenBerg age estimate |
| `Method` | Method used for the VandenBerg age determination |
| `HBtype` | Horizontal-branch morphology parameter |
| `R_G_VDB` | Galactocentric distance from VandenBerg |
| `M_V` | Integrated absolute V-band magnitude |
| `v_e0` | Central escape velocity |
| `RA` | Right ascension from Harris |
| `DEC` | Declination from Harris |
| `L` | Galactic longitude |
| `B` | Galactic latitude |
| `R_Sun` | Distance from the Sun |
| `R_gc_Harris` | Galactocentric distance from Harris |
| `X`, `Y`, `Z` | Galactic Cartesian coordinates |
| `v_r` | Radial velocity |
| `v_r_e` | Uncertainty in radial velocity |
| `v_LSR` | Velocity relative to the Local Standard of Rest |
| `sig_v` | Velocity dispersion |
| `sig_v_e` | Uncertainty in velocity dispersion |
| `c` | Cluster concentration parameter |
| `r_c` | Core radius |
| `r_h_Harris` | Half-light/half-radius quantity reported by Harris |
| `Mstar` | Stellar mass from Krause |
| `rh_Krause` | Half-mass radius from Krause |
| `C5` | Compactness parameter from Krause |
| `Age_Krause` | Cluster age from Krause |
| `FeH_Krause` | Metallicity from Krause |

## Cleaning Summary

- Original catalogue files were preserved unchanged in `data/raw/`.
- A common `Cluster_ID` was created by standardising identifier formatting.
- Catalogue-specific alternative names were resolved using documented aliases.
- Harris Parts I and III were merged using `Cluster_ID`.
- The VandenBerg catalogue was used as the primary sample, retaining all 55 clusters.
- Krause measurements were added where corresponding clusters were available.
- The final dataset contains 55 unique globular clusters with no duplicated `Cluster_ID` values.
- Three clusters in the VandenBerg sample do not have corresponding Krause measurements, so their Krause-derived quantities remain missing.

In [81]:
print("Final dataset shape:", cleaned.shape)
print("Unique clusters:", cleaned["Cluster_ID"].nunique())
print("Duplicate cluster IDs:", cleaned["Cluster_ID"].duplicated().sum())

Final dataset shape: (55, 38)
Unique clusters: 55
Duplicate cluster IDs: 0
